opentelemetry-contrib/src/trace/propagator/datadog.rs

```
@@ -0,0 +1,154 @@
+            .map(|id| TraceId::from_u128(id as u128))
+            .map_err(|_| ExtractError::InvalidTraceId)
+    }
+
+    fn extract_span_id(&self, span_id: &str) -> Result<SpanId, ExtractError> {
+        u64::from_str_radix(span_id, 10)
+            .map(SpanId::from_u64)
+            .map_err(|_| ExtractError::InvalidSpanId)
+    }
+
+    fn extract_sampling_priority(
+        &self,
+        sampling_priority: &str,
+    ) -> Result<SamplingPriority, ExtractError> {
+        let i = i32::from_str_radix(sampling_priority, 10)
+            .map_err(|_| ExtractError::InvalidSamplingPriority)?;
+
+        match i {
+            -1 => Ok(SamplingPriority::UserReject),
+            0 => Ok(SamplingPriority::AutoReject),
+            1 => Ok(SamplingPriority::AutoKeep),
+            2 => Ok(SamplingPriority::UserKeep),
+            _ => Err(ExtractError::InvalidSamplingPriority),
+        }
+    }
+
+    fn extract_span_context(&self, extractor: &dyn Extractor) -> Result<SpanContext, ExtractError> {
+        let trace_id =
+            self.extract_trace_id(extractor.get(DATADOG_TRACE_ID_HEADER).unwrap_or(""))?;
+        let span_id =
+            self.extract_span_id(extractor.get(DATADOG_PARENT_ID_HEADER).unwrap_or(""))?;
+        let sampling_priority = self.extract_sampling_priority(
+            extractor
+                .get(DATADOG_SAMPLING_PRIORITY_HEADER)
+                .unwrap_or(""),
+        )?;
+        let sampled = match sampling_priority {
+            SamplingPriority::UserReject | SamplingPriority::AutoReject => TRACE_FLAG_NOT_SAMPLED,
+            SamplingPriority::UserKeep | SamplingPriority::AutoKeep => TRACE_FLAG_SAMPLED,
+        };
+
+        let trace_state = TraceState::default();
+
+        let span_context = SpanContext::new(trace_id, span_id, sampled, true, trace_state);
+
+        if span_context.is_valid() {
+            Ok(span_context)
+        } else {
+            Err(ExtractError::InvalidSpanContext)
```

Could use[ global error handler](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry/src/global/error_handler.rs) to handle those errors as we cannot propagate it from `TextMapPropagator`

---

Do you meant use the Error enum in there instead?

---

Or I guess at least impl Into<error_handler::Error>, and then manually call `handle_error`?

---

Yeah we usually wrap the error into TraceError so users can choose how to handle them

---

I'm a bit hesitant to throw this into `handle_error` as I feel like this could be *VERY* high volume, and wouldn't want to have this just spam stderr for every trace that runs into this error. Exporter errors at least have the benefit of typically only happening once per batch interval where this would happen on every RPC call.

But I do think having this log *somehow* would be good instead of just silently eating the error. Not sure what the best move is here.

---

Good point. And currently, we don't have many logs in place, might need to look into that. I think we could leave it as it is and add structured logs once we have it